Explicitly list Packer plugin dependencies

jsf9k opened the pull request:
## 🗣 Description ##

This pull request explicitly adds the Packer plugin dependencies to the Packet template file.

> **Important**
> Note that this addition requires the user to either (1) use the `setup-env` script, or (2) run `packer init` or `packer init -upgrade` before `pre-commit` can run successfully.

## 💭 Motivation and context ##

* The syntax for this is supported as of Packer version 1.7.0, and eventually it will be required.  See [here](https://developer.hashicorp.com/packer/guides/1.7-template-upgrade) for more details.
* In any event, it is good to explicitly document dependencies.

There is [an issue for Dependabot to start tracking these kinds of dependencies](https://github.com/dependabot/dependabot-core/issues/8045).

## 🧪 Testing ##

All automated tests pass.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

jsf9k committed:
Explicitly add required plugins for Packer

packer init is supported as of Packer v1.7.0, and will eventually be
required.

See here for more details:
https://developer.hashicorp.com/packer/guides/1.7-template-upgrade

jsf9k committed:
Require Packer 1.7.0 or later

Packer 1.7.0 is the first version to support the required_plugins
syntax.

jsf9k committed:
Add tasks to install Packer plugins

These tasks must take place before either the Packer pre-commit linter
or packer build is run; otherwise, the packer executable will error
out.

jsf9k committed:
Add code to install Packer plugin dependencies

This must be done before pre-commit can run successfully.

mcdonnnj reviewed (commented):
I believe you can just run `packer init -upgrade src/` and it will capture all plugin requirements across any `.pkr.hcl` or `.pkr.json` files in the `src/` directory. As a reference if I try this in the [cisagov/cyhy_amis](https://github.com/cisagov/cyhy_amis) project (which notably has _neither_ type of file currently) I get:

```console
$ packer init -upgrade packer/
Error: Could not find any config file in packer/

A config file must be suffixed with `.pkr.hcl` or `.pkr.json`. A folder can be
referenced.
```

While running `packer init -upgrade src/` in this repository exits cleanly with no output.

jsf9k committed:
Simplify packer init command

Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>

jsf9k commented:
> I believe you can just run `packer init -upgrade src/` and it will capture all plugin requirements across any `.pkr.hcl` or `.pkr.json` files in the `src/` directory. As a reference if I try this in the [cisagov/cyhy_amis](https://github.com/cisagov/cyhy_amis) project (which notably has _neither_ type of file currently) I get:
> 
> ```
> $ packer init -upgrade packer/
> Error: Could not find any config file in packer/
> 
> A config file must be suffixed with `.pkr.hcl` or `.pkr.json`. A folder can be
> referenced.
> ```
> 
> While running `packer init -upgrade src/` in this repository exits cleanly with no output.

Please see commit 1595315.

dav3r reviewed (approved):
Good thinking. 👍 👍

mcdonnnj commented on src/packer.pkr.hcl:
```diff
@@ -1,3 +1,20 @@
+  # The required_plugins section is only supported in Packer 1.7.0 and
+  # later.  We also want to avoid jumping to Packer v2 until we are
+  # ready.
+  required_version = ">=1.7.0,<2"
```
This should work based on [the documentation](https://developer.hashicorp.com/packer/docs/templates/hcl_templates/blocks/packer#-3).

```suggestion
  required_version = "~> 1.7"
```

mcdonnnj commented on src/packer.pkr.hcl:
```diff
@@ -1,3 +1,20 @@
+  required_plugins {
+    amazon = {
+      source  = "github.com/hashicorp/amazon"
+      version = "~> 1"
```
I am concerned that this will allow installation of the next major version (which I do not believe we would like to happen). The below matches the current major.minor for this plugin.

```suggestion
      version = "~> 1.2"
```

mcdonnnj commented on src/packer.pkr.hcl:
```diff
@@ -1,3 +1,20 @@
+    }
+    ansible = {
+      source  = "github.com/hashicorp/ansible"
+      version = "~> 1"
```
I am concerned that this will allow installation of the next major version (which I do not believe we would like to happen). The below matches the current major.minor for this plugin.

```suggestion
      version = "~> 1.1"
```

mcdonnnj reviewed (changes requested):
Apologies I completely missed the plugin versioning on my first skim so this is belated.

jsf9k committed:
Simplify version constraint for Packer itself

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Correct version constraints

The version constraints should allow updates up to but not including the next major version of the plugin.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj reviewed (approved):
LGTM ✔
